src/lib/storage/dictionary_segment.hpp

```diff
@@ -0,0 +1,154 @@
+    } else {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint32_t>>(base_segment->size());
+    }
+
+    // Build up the attribute vector by looking up the index of the value in the dictionary for each value.
+    // Because the dictionary is already sorted, we can use the efficient std::lower_bound method for performing a
+    // binary search for the dictionary index.
+    for (size_t value_index = (ChunkOffset)0; value_index < base_segment->size(); value_index++) {
+      uint32_t dictionary_index =
+          std::distance(_dictionary->begin(), std::lower_bound(_dictionary->begin(), _dictionary->end(),
+                                                               type_cast<T>(base_segment->operator[](value_index))));
+      _attribute_vector->set(value_index, ValueID{dictionary_index});
+    }
+  }
+
+  // SEMINAR INFORMATION: Since most of these methods depend on the template parameter, you will have to implement
+  // the DictionarySegment in this file. Replace the method signatures with actual implementations.
+
+  // return the value at a certain position. If you want to write efficient operators, back off!
+  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override {
+    return _dictionary->at(_attribute_vector->get(chunk_offset));
+  }
+
+  // return the value at a certain position.
+  T get(const size_t chunk_offset) const { return _dictionary->at(_attribute_vector->get(chunk_offset)); }
+
+  // dictionary segments are immutable
+  void append(const AllTypeVariant&) override { throw std::exception(); }
+
+  // returns an underlying dictionary
+  std::shared_ptr<const std::vector<T>> dictionary() { return _dictionary; }
+
+  // returns an underlying data structure
+  std::shared_ptr<const BaseAttributeVector> attribute_vector() { return _attribute_vector; }
+
+  // return the value represented by a given ValueID
+  const T& value_by_value_id(ValueID value_id) { return _dictionary->at(value_id); }
+
+  // returns the first value ID that refers to a value >= the search value
+  // returns INVALID_VALUE_ID if all values are smaller than the search value
+  ValueID lower_bound(T value) const {
+    for (size_t value_index = 0; value_index < _attribute_vector->size(); value_index++) {
+      if (get(value_index) >= value) {
+        return ValueID{value_index};
+      }
+    }
+    return INVALID_VALUE_ID;
+  }
+
+  // same as lower_bound(T), but accepts an AllTypeVariant
+  ValueID lower_bound(const AllTypeVariant& value) const { return lower_bound(type_cast<T>(value)); }
+
+  // returns the first value ID that refers to a value > the search value
+  // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
+  ValueID upper_bound(T value) const {
+    for (size_t value_index = 0; value_index < _attribute_vector->size(); value_index++) {
+      if (get(value_index) > value) {
+        return ValueID{value_index};
+      }
+    }
+    return INVALID_VALUE_ID;
```

Comment:
You could also use std::upper_bound

Reply:
resolved in 7827984eacf7c7a71dab1a93f953bb16094efbb0